app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt

```diff
@@ -0,0 +1,112 @@
+                    MovieCategory.POPULAR_ON_TV -> movieService.fetchPopularMovies()
+                    MovieCategory.POPULAR_FOR_RENT -> movieService.fetchNowPlayingMovies()
+                    MovieCategory.POPULAR_IN_THEATERS -> movieService.fetchTopRatedMovies()
+                    MovieCategory.POPULAR_STREAMING -> movieService.fetchUpcomingMovies()
+                    MovieCategory.NOW_PLAYING_MOVIES -> movieService.fetchNowPlayingMovies()
+                    MovieCategory.NOW_PLAYING_TV -> movieService.fetchNowPlayingMovies()
+                    MovieCategory.UPCOMING_TODAY -> movieService.fetchUpcomingMovies()
+                    MovieCategory.UPCOMING_THIS_WEEK -> movieService.fetchNowPlayingMovies()
+                }
+                emit(movieResponse.movies)
+            }.flatMapLatest { apiMovies ->
+                movieDao.favorites()
+                    .map { favoriteMovies ->
+                        apiMovies.map { apiMovie ->
+                            apiMovie.toMovie(isFavorite = favoriteMovies.any { it.id == apiMovie.id })
+                        }
+                    }
+            }.shareIn(
+                scope = CoroutineScope(bgDispatcher),
+                started = SharingStarted.WhileSubscribed(1000L),
+                replay = 1,
+            )
+        }
+
+    private val favorites = movieDao.favorites().map {
+        it.map { dbFavoriteMovie ->
+            Movie(
+                id = dbFavoriteMovie.id,
+                imageUrl = dbFavoriteMovie.posterUrl,
+                title = "",
+                overview = "",
+                isFavorite = true,
+            )
+        }
+    }.shareIn(
+        scope = CoroutineScope(bgDispatcher),
+        started = SharingStarted.WhileSubscribed(1000L),
+        replay = 1,
+    )
+
+    override fun movies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    override fun movieDetails(movieId: Int): Flow<MovieDetails> = flow {
+        emit(movieService.fetchMovieDetails(movieId) to movieService.fetchMovieCredits(movieId))
+    }.flatMapLatest { (apiMovieDetails, apiMovieCredits) ->
+        movieDao.favorites()
+            .map { favoriteMovies ->
+                apiMovieDetails.toMovieDetails(
+                    isFavorite = favoriteMovies.any { it.id == apiMovieDetails.id },
+                    crew = apiMovieCredits.crew,
+                    cast = apiMovieCredits.cast
+                )
+            }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    override suspend fun addMovieToFavorites(movieId: Int) {
+        val movie = findMovie(movieId)
+        movie?.imageUrl?.let { movieDao.insertMovie(DbFavoriteMovie(movie.id, it)) }
+    }
+
+    override suspend fun removeMovieFromFavorites(movieId: Int) = movieDao.deleteMovie(movieId)
+
+    override suspend fun toggleFavorite(movieId: Int) {
+        runBlocking(bgDispatcher) {
```

Comment:
We [shouldn't](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/run-blocking.html) be using `runBlocking` function here. Instead, you can switch context like this:
```kotlin
withContext(bgDispatcher) {

}
```